zebra/zebra_srte.c

```diff
@@ -195,19 +248,100 @@ static void zebra_sr_policy_notify_update(struct zebra_sr_policy *policy)
 	}
 
 	rnh = zebra_lookup_rnh(&p, zvrf_id(zvrf), SAFI_UNICAST);
-	if (!rnh)
+	if (rnh == NULL)
 		return;
 
+	if (policy->type == ZEBRA_SR_POLICY_TYPE_SRV6) {
+		/* check color */
+		for (; rnh; rnh = rnh->next)
+			if (rnh->srte_color == policy->color)
+				break;
+		if (rnh == NULL)
+			return;
+		rnh->srp_status = policy->status;
+	}
+
+	if (IS_ZEBRA_DEBUG_SRV6) {
+		char endpoint[IPADDR_STRING_SIZE];
+
+		ipaddr2str(&policy->endpoint, endpoint, sizeof(endpoint));
+		zlog_debug("%s: endpoint %s color %u status %u",
+			__func__, endpoint, policy->color, policy->status);
+	}
+
 	for (ALL_LIST_ELEMENTS_RO(rnh->client_list, node, client)) {
 		if (policy->status == ZEBRA_SR_POLICY_UP)
 			zebra_sr_policy_notify_update_client(policy, client);
-		else
-			/* Fallback to the IGP shortest path. */
-			zebra_send_rnh_update(rnh, client, zvrf_id(zvrf),
-					      policy->color);
+		else {
+			if (policy->type == ZEBRA_SR_POLICY_TYPE_SRV6)
+				zebra_sr_policy_notify_unknown(rnh, client);
+			else
+				/* Fallback to the IGP shortest path. */
+				zebra_send_rnh_update(rnh, client, zvrf_id(zvrf),
+								policy->color);
+		}
 	}
 }
 
+int zebra_sr_policy_notify_unknown(struct rnh *rnh,
+						struct zserv *client)
+{
+	struct stream *s;
+	uint32_t message = 0;
+	struct route_node *rn;
+
+	rn = rnh->node;
+
+	/* Get output stream. */
+	s = stream_new(ZEBRA_MAX_PACKET_SIZ);
+
+	zclient_create_header(s, ZEBRA_NEXTHOP_UPDATE, rnh->vrf_id);
+
+	/* Message flags. */
+	SET_FLAG(message, ZAPI_MESSAGE_SRTE);
+	stream_putl(s, message);
+	stream_putw(s, rnh->safi);
+
+	switch (rn->p.family) {
+	case AF_INET:
+		stream_putw(s, AF_INET);
+		stream_putc(s, IPV4_MAX_BITLEN);
+		stream_put_in_addr(s, &rn->p.u.prefix4);
+		stream_putw(s, AF_INET);
+		stream_putc(s, IPV4_MAX_BITLEN);
+		stream_put_in_addr(s, &rn->p.u.prefix4);
+		break;
+	case AF_INET6:
+		stream_putw(s, AF_INET6);
+		stream_putc(s, IPV6_MAX_BITLEN);
+		stream_put(s, &rn->p.u.prefix6, IPV6_MAX_BYTELEN);
+		stream_putw(s, AF_INET6);
+		stream_putc(s, IPV6_MAX_BITLEN);
+		stream_put(s, &rn->p.u.prefix6, IPV6_MAX_BYTELEN);
+		break;
+	default:
+		flog_warn(EC_LIB_DEVELOPMENT,
+			  "%s: unknown policy endpoint address family: %u",
+			  __func__, rn->p.family);
+		exit(1);
+	}
+
+	stream_putl(s, rnh->srte_color);
+
```

Comment:
Not sure it is correct. If a colored nht is not available, then by default the colorless nht is to be used.
What is done here is: there is no path for that colored nexthop, path is invalid.

 if BGP reads this, it will understand: I will withdraw my BGP update.

My expectation is that BGP should eventually consider that the colored nht is not available, BUT it will reuse the colorless  NHT available. I can not say if this code is correct or not, because the BGP code associated is not present.